lib/take_elevated.lua

```diff
@@ -0,0 +1,98 @@
+local tokens_per_ms               = tonumber(ARGV[1])
+local bucket_size                 = tonumber(ARGV[2])
+local tokens_to_take              = tonumber(ARGV[3])
+local ttl                         = tonumber(ARGV[4])
+local drip_interval               = tonumber(ARGV[5])
+local erl_tokens_per_ms           = tonumber(ARGV[6])
+local erl_bucket_size             = tonumber(ARGV[7])
+local erl_activation_period_seconds  = tonumber(ARGV[8])
+
+-- the key to use for pulling last bucket state from redis
+local lastBucketStateKey = KEYS[1]
+
+-- the key for checking in redis if elevated rate limits (erl) were activated earlier
+local erlKey = KEYS[2]
+local is_erl_activated = redis.call('EXISTS', erlKey)
+
+-- get current bucket state
+local current = redis.pcall('HMGET', lastBucketStateKey, 'd', 'r')
+if current.err ~= nil then
+    current = {}
+end
+
+-- get current time from redis, to be used in new bucket size calculations later
+local current_time = redis.call('TIME')
+local current_timestamp_ms = current_time[1] * 1000 + current_time[2] / 1000
+
+local function calculateNewBucketContent(current, tokens_per_ms, bucket_size, current_timestamp_ms)
+    if current[1] and tokens_per_ms then
+        -- drip bucket
+        local last_drip = current[1]
+        local content = current[2]
+        local delta_ms = math.max(current_timestamp_ms - last_drip, 0)
+        local drip_amount = delta_ms * tokens_per_ms
+        return math.min(content + drip_amount, bucket_size)
+    elseif current[1] and tokens_per_ms == 0 then
+        -- fixed bucket
+        return current[2]
+    else
+        -- first take of the bucket
+        return bucket_size
+    end
+end
+
+-- Enable verbatim replication to ensure redis sends script's source code to all masters
+-- managing the sharded database in a clustered deployment.
+-- https://redis.io/docs/interact/programmability/eval-intro/#:~:text=scripts%20debugger.-,Script%20replication,-In%20standalone%20deployments
+redis.replicate_commands()
+
+-- calculate new bucket content
+local bucket_content_after_refill
+if is_erl_activated==1 then
+    bucket_content_after_refill = calculateNewBucketContent(current, erl_tokens_per_ms, erl_bucket_size, current_timestamp_ms)
+else
+    bucket_content_after_refill = calculateNewBucketContent(current, tokens_per_ms, bucket_size, current_timestamp_ms)
+end
+
+local enough_tokens = bucket_content_after_refill >= tokens_to_take
+local bucket_content_after_take = bucket_content_after_refill
+
+if enough_tokens then
+    if is_erl_activated == 1 then
+        bucket_content_after_take = math.min(bucket_content_after_refill - tokens_to_take, erl_bucket_size)
+    else
+        bucket_content_after_take = math.min(bucket_content_after_refill - tokens_to_take, bucket_size)
+    end
+else
+    -- if tokens are not enough, see if activating erl will help.
+    if is_erl_activated == 0 then
```

Comment:
No, however it shouldn't be a blocker to merge this change. Mind that this new `lua` script is only run when ERL is explicitly activated - otherwise, we use the exact same old flow, and the old lua script. We should run the load tests before activating the feature though.

That being said - do we have any baseline for expected performance? Was the previous implementation load-tested? Would be useful if we already have it